Add source, creator and tag links to the single result page media info

obulat opened the pull request:
<!-- prettier-ignore -->
## Fixes
<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, please consider opening one before creating this pull request. -->

Fixes #2730 by @obulat

## Description
<!-- Concisely describe what the pull request does. -->
<!-- Add screenshots, videos, or other media to show the problem and the solution when appropriate. -->
This PR updates the single result page's creator, source and tag links to the new designs, linking them to the collection pages. 

### Audio
<img width="840" alt="Screenshot 2023-11-13 at 12 33 00 PM" src="https://github.com/WordPress/openverse/assets/15233243/dc96ae3f-6d04-4a3e-ba3a-052db6e09407">

### Image 
<img width="842" alt="Screenshot 2023-11-13 at 12 34 38 PM" src="https://github.com/WordPress/openverse/assets/15233243/f9bd70cf-0848-4ed3-8d79-8003cd8da3ce">

If the creator and source names are longer than the page width, they become scrollable. Setting up the scrolling to work in both LTR and RTL was a bit complex.

Note that the link URLs will be updated after #3140 is merged because it's impossible to add a link to a localized page that does not exist (and we don't currently have pages for `source`/`creator`/`tag` collections).

## Testing Instructions
<!-- Give steps for the reviewer to verify that this PR fixes the problem; or delete this section entirely. -->
Run the app
Go to localhost:8443/preferences and switch the `additional_search_views` flag `on`
Go to single result pages for audio and image. 
The source, creator and tag should have an updated style with correct links (`/image/tag/cat`, `/image/source/flickr`, `/image/source/flickr/creator/name`). Note that i18n (linking to the correct locale if you are not on the English one) does not work yet because the pages for tag/source/creator do not exist yet. The links also do not open the pages because the pages are created in #3140 
When the creator and the source name are longer than the available width, they become scrollable. To check, see http://localhost:8443/image/8c8f2318-acc8-477c-a059-34f44901933e?q=cat

## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. -->

- [ ] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [ ] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [ ] My commit messages follow [best practices][best_practices].
- [ ] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no visible errors.
- [ ] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

fcoveram reviewed (changes requested):
Besides the comment shared, I noticed two more things.

* The column and row distance between tag items should be 12px instead of 8px. [Here is](https://share.cleanshot.com/yH0YdVwz) what I see.
* In `xs`, the author and source area don't have the gradient and spacing applied as in the design. [Here is](https://share.cleanshot.com/WRqncVvg) what I see. Below you can find the specs for both elements.

### Gradient
width: 64px; height: 32px. The white fill goes from 100% to 0% opacity.

<img src="https://github.com/WordPress/openverse/assets/895819/0615ce01-b134-4fdf-8416-4eb2c734aa63" width="300">

### Spacing
12px between filter buttons and left and right arrows

<img src="https://github.com/WordPress/openverse/assets/895819/35660029-810b-4643-8b5e-aa53748234fc" width="300">

obulat commented on frontend/src/components/VAudioTrack/layouts/VFullLayout.vue:
```diff
@@ -16,6 +16,21 @@
     </div>
     <div
+      v-if="additionalSearchViews"
+      class="mx-auto grid grid-cols-1 grid-rows-[auto,1fr] gap-y-6 p-6 pb-0 sm:grid-cols-[1fr,auto] sm:grid-rows-1 sm:gap-x-6 lg:mb-6 lg:max-w-5xl"
```
After discussing this synchronously, we decided to add a top margin of 4px to the CTA (Get media button) to make it well-balance both for single line titles and longer titles.

fcoveram reviewed (approved):
The spacing changes in `sm` and beyond are correct, and the scrolling interaction works well.

The only remaining points in `xs` are:
* Gradient is still not fully white on the sides. As in [this screenshot](https://share.cleanshot.com/2My3MxzW).
* Spacing between left and right action and filter buttons is still tight. As in [this screenshot](https://share.cleanshot.com/LM9Myflt).
* Shaking effect in the pill when the scroll reaches its end. As in [this screenshot](https://share.cleanshot.com/mldCvWSD).
* Left and right buttons don't show up immediately. As in [this screenshot](https://share.cleanshot.com/pt0rgtHp).

I can create a new ticket for that as this PR addresses the main changes and the above is more subtle.

sarayourfriend reviewed (changes requested):
I'm not sure if the shaking that Francisco refers to in his review is what I'm seeing in the video below, but I think it presents a major usability issue that looks really easy to run into.

[Screencast_20231122_095132.webm](https://github.com/WordPress/openverse/assets/24264157/4ba3bb16-8674-4d68-a11f-62fb2221e452)

The scenario is basically: I want to scroll all the way to the end of the provider to read the provider name. To do this, I'm repeatedly and quickly clicking the arrow to scroll the view. However, because the arrow disappears at the end and the region I was clicking is now the provider link, I'm suddenly taken to a completely different page. When using normal scrolling arrows, my experience has always been that once you reach the end, the arrow just stops working, rather than disappearing.

The jittering that I think Francisco is also referring to is really jarring if you're trying to read the scrolling text, because the text moves rapidly twice, once to make room for the arrows and then again to actually start scrolling. It'd be one thing if they ended up in the same location, which would at least not cause the jarring effect, but it's ever so slightly different, which means your eyes end up tracking the entire movement and noticing it more.

I don't know how to fix the disappearing arrows usability issue. It feels unavoidable given the design of the arrows appearing. The jittering though, looks like it could be solved by rendering the arrows "above" the scrolling pills, rather than next to them, so that the browser doesn't try to move the contents of the element to make room for the arrows, with the JS update to scroll the contents based on the arrow click coming after the browser paint.

I'm not leaving an approve/request changes on this because I'm not confident about either, so just commenting. If it's not possible to fix these problems in this PR, or if it's better to fix them in a follow up, then I can approve, but on the condition that the feature rollout is blocked on fixing at least one of these issues. The usability problem feels important to me because I think a lot of users will run into it and find it frustrating, but the second is a visual accessibility issue, in my opinion. I _think_ that fixing the first problem would naturally fix the second, which is why I say only one of them needs to be fixed. Really the second one "needs" to be fixed, and if we can fix the first one too it would be a significant improvement to the user experience.

While writing this review and doing more testing, I noticed something I do need to request changes for, however. There is also a big accessibility issue for sighted keyboard users:

![image](https://github.com/WordPress/openverse/assets/24264157/2bf38675-fd42-41d4-9449-b4b4c1f4dac5)

In that screenshot the author button is focused. There is a tiny sliver of the purple outline, but otherwise there's no indication that it is focused and, worse, the button is focused but not in view, so I can't even read what it says. Same with the source button. If I tab to the next button, this is what it looks like:

![image](https://github.com/WordPress/openverse/assets/24264157/b38bd658-72bd-48c7-8c6a-1f22b36c6bf7)

It's almost indistinguishable from the focus state of the previous button and again, I'm unable to read the contents of the button.

My suggestion for fixing this needs @fcoveram's review but is as follows:

1. For the focus ring issue, either use a different variant that changes the colour of the button when focused or increase the height of the container so that the top and bottom of the focus ring is visible.

![image](https://github.com/WordPress/openverse/assets/24264157/28a5724b-b0c5-47a8-a609-4d8ef74cae55)

In that screenshot I increased it to 36 px up from 32 and it still doesn't show both the top and the bottom, but it's clearly an improvement. I used taller heights as well but it looks like the buttons need vertical justification within the container as well, because it only adds space to the bottom of the buttons rather than around them. Alternatively, maybe all buttons actually need additional margin to account for the size of the focus ring? When I add margin of 4px to the button I can see the focus ring without making any many changes to the height of the container. That also avoids the vertical justification issue.
2. For the button contents not being visible, automatically scroll the button completely into view when it's focused. I think a non-hijacked scrolling view would do this automatically, so I guess we need to reimplement it?

obulat commented on frontend/src/components/VMediaInfo/VByLine/VByLine.vue:
```diff
@@ -0,0 +1,301 @@
+
+    const buttonVariant = computed(() =>
+      uiStore.isBreakpoint("sm") ? "transparent-gray" : "filled-gray"
+    )
```
I used a single button variant, and added tailwind classes with and without `sm:` prefix to override the colors. Now, the change is much smoother.

sarayourfriend commented:
> Youtube tags also have the same behavior

Comparing our implementation with the youtube one, I think the youtube tags don't have as much of a jarring experience to me for two reasons:

1. There isn't the jitter of the arrows causing the tags to jump around
2. The arrow disappears immediately after clicking when it would scroll to the end. In other words, it doesn't scroll to the end and _then_ hide the arrow. It hides the arrow before it hits the end, after you've clicked the last possible "scroll".
3. It scrolls much faster and further, and feels more responsive, so I'm less inclined to rapidly click the scroll button to get to the end.

A lot of that is definitely improved a lot now though. The only weird thing I'm noticing is that on `image/8c8f2318-acc8-477c-a059-34f44901933e` I have to click twice on the right scroll to get to the end, but once I'm at the end, only once on the left scroll to get back to the start :thinking: I don't know what could be causing that.

Maybe we could use real-time calculation of scroll distance? Like, scroll at least 2/3 the width of the container every time, but fudge that a bit so that we're never scrolling less than 1/3? I don't know exactly how to make that math work out, but the idea would be that you'd never have a scroll button press that only scrolled a tiny amount, which is what's happening with the second click of the right scroll for me on that result. Here's a video showing that.

[Screencast_20231123_121008.webm](https://github.com/WordPress/openverse/assets/24264157/6ee18dea-0934-4963-b113-efc514183a9a)


Probably not blocking, but I haven't fully re-reviewed now. The improvements are awesome though, for sure. I'll re-review tomorrow morning (logging off for the day now).

fcoveram commented:
Great comments here. Three topics have been touched, so grouping them to share my thoughts.

## Filters interaction

Thanks @sarayourfriend for sharing your use case. The reason behind using arrows (visible or hidden when reaching the end) comes from an a11y assumption of providing controls, and because some filters' text lengths are quite long and need to be contained within the area.

I revisited YouTube again and you're right about the behavior. But I also noticed that buttons are hidden on mobile. It might be because touch interaction predominates in small devices and the scroll relies on swiping, so removing the arrows on small breakpoints could be a good idea.

If we can test your suggestion in a different ticket, it would be great to decide which interaction feels smoother. So far, I would merge this PR as it is.

## Testing the implementation

> I also experienced this (and got really annoyed with it) when developing this locally.

I'd like to reinforce the idea that design work is not written in stone. If you experience something that feels weird, unexpected, or literally breaks up your experience as a user, please share it. I've been contributing to other WordPress projects and noticed that we don't tend to discuss design ideas and instead develop what's on mockups.

We all are Openverse users and our experiences with the tool provide tremendous value.

## Visible focus

> There is a tiny sliver of the purple outline, but otherwise there's no indication that it is focused and, worse, the button is focused but not in view

I've mentioned this problem in other tickets and raising the question again of whether is possible to address it for the whole site. If we continue revisiting the focus style per case, we risk missing future use cases like this one, where I completely forgot about it.

We can move the discussion into a different ticket, indeed. But to me, this component style belongs to a core state that should not be in revision because of possible layout breaks.

obulat commented:
> I'd like to reinforce the idea that design work is not written in stone. If you experience something that feels weird, unexpected, or literally breaks up your experience as a user, please share it. I've been contributing to other WordPress projects and noticed that we don't tend to discuss design ideas and instead develop what's on mockups.

> We all are Openverse users and our experiences with the tool provide tremendous value.

Thank you for reminding about this, @fcoveram! You are right that we should all be more mindful about the UX when working on the frontend issues.

> We can move the discussion into a different ticket, indeed. But to me, this component style belongs to a core state that should not be in revision because of possible layout breaks.

The core state **does** have the focus state implemented. However, when an element or its container has a CSS position set, the focus ring gets cut off. It still exists, but it's hidden behind other elements, so it becomes useless. I think it would be nice to open an issue for a better fix for this problem than what I implemented here (adding a padding to the container and removing some of the margins to balance the height with the mockup.

fcoveram commented:
> The core state does have the focus state implemented.

You are right, I didn't explain myself correctly. It is part of the component.

> I think it would be nice to open an issue for a better fix for this problem…

Mainly to suggest ideas and land on a solution like adding extra space as we do it but from the design stage or another one I'm unaware of.

obulat commented:
> The only weird thing I'm noticing is that on image/8c8f2318-acc8-477c-a059-34f44901933e I have to click twice on the right scroll to get to the end, but once I'm at the end, only once on the left scroll to get back to the start 🤔 I don't know what could be causing that.

I think this was caused by 2 things. The issue has suggested using `scroll-snap` for scrolling. When I added it, it was causing a problem when the buttons were about as long as the scroll width: you could not scroll past them.

[Scroll snap problem.webm](https://github.com/WordPress/openverse/assets/15233243/3ff57c07-11cb-4254-bc69-c1e0e6a355c1)

I removed the `scroll snap` and also adjusted scrolling for the last two steps: when the distance left is less than 2 scroll steps, the scroll step becomes half of this distance. So, if the distance left is 200, we don't scroll 150px and 50px (scroll step and what's left), but scroll 100px and 100px instead.

Here's the updated scrolling:
[Scrolling.webm](https://github.com/WordPress/openverse/assets/15233243/d3cf6968-10ba-430a-9c77-afc54087b466)

> It scrolls much faster and further, and feels more responsive, so I'm less inclined to rapidly click the scroll button to get to the end.

@sarayourfriend, do you think it would make sense to increase the scroll step? I took 150px as "about half of the screen on mobile width", but it might be better to have larger steps?

obulat commented:
> I revisited YouTube again and you're right about the behavior. But I also noticed that buttons are hidden on mobile. It might be because touch interaction predominates in small devices and the scroll relies on swiping, so removing the arrows on small breakpoints could be a good idea.

I looked at Facebook and Google maps on mobile, and neither of them has arrows for their horizontally scrollable list of tags. I guess this is the behavior users expect, so I would vote for removing the arrows on mobile.

> If we can test your suggestion in a different ticket, it would be great to decide which interaction feels smoother. So far, I would merge this PR as it is.

Merging this PR would make it easier to test the PR that adds collection pages, so I would be happy to do as you suggest. @sarayourfriend , do you think this is okay as is, considering that it's under a feature flag and we can open new issues to keep iterating?

sarayourfriend reviewed (approved):
@obulat Yes, go for it! I intended to leave an approving review on Friday with that exact stipulation. Any of the outstanding issues can be resolved in smaller follow-up PRs, that would be overall easier to review anyway.

I also wanted to say good work on the scrolling box. It is complicated, and I wouldn't have known where to start with it, personally. Despite the small issues, it's a really excellent first pass at it that wouldn't even be bad to release as public, if we were tight on time. The various issues are rather minor and don't overall impede the usability (now that the major a11y focus issue is solved).

For me the issues to resolve are:

1. Adjusting the scroll distance to make it feel good (fast enough)
2. Potentially removing the arrows on mobile (making sure it's visually clear that there is a scrollable area there, e.g., with the fading)
3. Scrolling buttons 100% into view when they're focused via keyboard
